server/src/main/java/org/apache/druid/server/http/DataSourcesResource.java

```diff
@@ -700,4 +762,36 @@ static boolean isSegmentLoaded(Iterable<ImmutableSegmentLoadInfo> serverView, Se
     }
     return false;
   }
+
+  @VisibleForTesting
+  protected static class MarkDatasourceSegmentsPayload
+  {
+    private final Interval interval;
+    private final Set<String> segmentIds;
+
+    @JsonCreator
+    public MarkDatasourceSegmentsPayload(
+        @JsonProperty("interval") Interval interval,
+        @JsonProperty("segmentIds") Set<String> segmentIds
+    )
+    {
+      this.interval = interval;
+      this.segmentIds = segmentIds;
+    }
+
+    public Interval getInterval()
+    {
+      return interval;
+    }
+
+    public Set<String> getSegmentIds()
+    {
+      return segmentIds;
+    }
+
+    public boolean isValid()
```

Comment:
This could just be a preconditions check in the constructor, and I would use a different error message that just tells the user that `intervals` and `segmentIds` cannot both be set.

The current error message I don't think is accurate ("Request payload contains invalid interval[%s] or segmentIds[%s], atmost one valid value must be provided"), since the check is not really for validity of the supplied values but for the presence of both.

Reply:
Yeah can move as a preconditions check in constructor and change the message.

Reply:
oh, actually, if I check in preconditions, the API throws an `IllegalArgumentException`, but what we want is to return `Response.Status.BAD_REQUEST`

Reply:
ah, that's fine then